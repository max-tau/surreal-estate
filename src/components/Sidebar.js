import React, { useState } from "react";
import { Link, useLocation, useNavigate } from "react-router-dom";
import qs from "qs";
import "../styles/sidebar.css";

const Sidebar = () => {
  const [searchTerm, setSearchTerm] = useState("");
  const { search } = useLocation();

  const buildQueryString = (operation, valueObj) => {
    const currentQueryParams = qs.parse(search, { ignoreQueryPrefix: true });
    const newQueryParams = {
      ...currentQueryParams,
      [operation]: JSON.stringify(valueObj),
    };
    return qs.stringify(newQueryParams, {
      addQueryPrefix: true,
      encode: false,
    });
  };

  const navigate = useNavigate();

  const handleSearch = (event) => {
    event.preventDefault();
    const newQueryString = buildQueryString("query", {
      title: { $regex: `${searchTerm}` },
    });
    navigate(newQueryString);
  };
  const handleInputChange = (event) => {
    setSearchTerm(event.target.value);
  };

  return (
    <div className="sidebar">
      <form className="sidebar-search" onSubmit={handleSearch}>
        <input
          type="text"
          className="sidebar-search-input"
          onChange={handleInputChange}
          value={searchTerm}
        />
        <button className="sidebar-search-button" type="submit">
          <img
            className="search-icon"
            src="https://cdn-icons-png.flaticon.com/128/3308/3308543.png"
            alt="search icon"
          />
        </button>
      </form>
      <h3 className="sidebar-heading">Filter by city</h3>
      <ul className="sidebar-links">
        <li>
          <Link
            className="sidebar-links-item"
            to={buildQueryString("query", { city: "Manchester" })}
          >
            Manchester
          </Link>
        </li>
        <li>
          <Link
            className="sidebar-links-item"
            to={buildQueryString("query", { city: "Leeds" })}
          >
            Leeds
          </Link>
        </li>
        <li>
          <Link
            className="sidebar-links-item"
            to={buildQueryString("query", { city: "Sheffield" })}
          >
            Sheffield
          </Link>
        </li>
        <li>
          <Link
            className="sidebar-links-item"
            to={buildQueryString("query", { city: "Liverpool" })}
          >
            Liverpool
          </Link>
        </li>
      </ul>
      <h3 className="sidebar-heading">Sort By</h3>
      <ul className="sidebar-links">
        <li>
          <Link
            className="sidebar-links-item"
            to={buildQueryString("sort", { price: 1 })}
          >
            Price ascending
          </Link>
        </li>
        <li>
          <Link
            className="sidebar-links-item"
            to={buildQueryString("sort", { price: -1 })}
          >
            Price descending
          </Link>
        </li>
      </ul>
    </div>
  );
};

export default Sidebar;
