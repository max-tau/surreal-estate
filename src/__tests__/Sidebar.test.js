import React from "react";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import Sidebar from "../components/Sidebar";

describe("Sidebar", () => {
  it("renders components correctly", () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Sidebar />
      </BrowserRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
