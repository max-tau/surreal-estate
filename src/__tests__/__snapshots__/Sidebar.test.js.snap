// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar renders components correctly 1`] = `
<DocumentFragment>
  <div
    class="sidebar"
  >
    <form
      class="sidebar-search"
    >
      <input
        class="sidebar-search-input"
        type="text"
        value=""
      />
      <button
        class="sidebar-search-button"
        type="submit"
      >
        <img
          alt="search icon"
          class="search-icon"
          src="https://cdn-icons-png.flaticon.com/128/3308/3308543.png"
        />
      </button>
    </form>
    <h3
      class="sidebar-heading"
    >
      Filter by city
    </h3>
    <ul
      class="sidebar-links"
    >
      <li>
        <a
          class="sidebar-links-item"
          href="/?query={\\"city\\":\\"Manchester\\"}"
        >
          Manchester
        </a>
      </li>
      <li>
        <a
          class="sidebar-links-item"
          href="/?query={\\"city\\":\\"Leeds\\"}"
        >
          Leeds
        </a>
      </li>
      <li>
        <a
          class="sidebar-links-item"
          href="/?query={\\"city\\":\\"Sheffield\\"}"
        >
          Sheffield
        </a>
      </li>
      <li>
        <a
          class="sidebar-links-item"
          href="/?query={\\"city\\":\\"Liverpool\\"}"
        >
          Liverpool
        </a>
      </li>
    </ul>
    <h3
      class="sidebar-heading"
    >
      Sort By
    </h3>
    <ul
      class="sidebar-links"
    >
      <li>
        <a
          class="sidebar-links-item"
          href="/?sort={\\"price\\":1}"
        >
          Price ascending
        </a>
      </li>
      <li>
        <a
          class="sidebar-links-item"
          href="/?sort={\\"price\\":-1}"
        >
          Price descending
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
